src/backends/kimchi/prover.rs

```diff
@@ -233,3 +233,59 @@ impl VerifierIndex {
         .wrap_err("kimchi: failed to verify the proof")
     }
 }
+
+mod testing {
+    use crate::{
+        backends::{
+            kimchi::{KimchiVesta, VestaField},
+            Backend,
+        },
+        constants::Span,
+        var::{ConstOrCell, Value, Var},
+        witness::WitnessEnv,
+    };
+
+    #[test]
+    fn test_public_output_constraint() -> miette::Result<()> {
+        // setup a simple circuit
+        let mut backend = KimchiVesta::new(false);
+
+        let span = Span::default();
+        let input_val = VestaField::from(1);
+        let input_var = backend.add_public_input(Value::Constant(input_val), span);
+
+        let output_var = backend.add_public_output(Value::PublicOutput(None), span);
+
+        // 1 + 1 = 2
+        let const_val = VestaField::from(1);
+        let res_val = input_val + const_val;
+        let res_var = backend.add_const(&input_var, &const_val, span);
+        let public_outputs = Some(Var::new_cvar(ConstOrCell::Cell(output_var), span));
+        let returned_cells = Some(Vec::from([res_var]));
```

Comment:
I think it would have been nicer to compile something like this instead to make sure the end-to-end flow doesn't have a bug (and then it's easier to check any backend with it as well):

```rust
fn main(pub public_input: Field, private_input: Field) -> Field {
    let xx = private_input + public_input;
    assert_eq(xx, 2);
    let yy = xx + 6;
    return yy;
}
```

the second thing is that... the verify function doesn't check the permutation, so it's not going to detect the bug (if there is one) :D I believe we actually have to create a proof

Reply:
so something like that, roughly:

```rust
let code = r#"fn main(pub public_input: Field, private_input: Field) -> Field {
    let xx = private_input + public_input;
    assert_eq(xx, 2);
    let yy = xx + 6;
    return yy;
}";

let mut sources = Sources::new();
let mut tast = TypeChecker::new();
let this_module = None;
let _node_id = typecheck_next_file(
    &mut tast,
    this_module,
    &mut sources,
    "public_output.no".to_string(),
    code.clone(),
    0,
)
.unwrap();

let kimchi_vesta = KimchiVesta::new(false);
let compiled_circuit = compile(&sources, tast, kimchi_vesta)?;

let (prover_index, verifier_index) = compiled_circuit.compile_to_indexes().unwrap();

let generated_witness = generate_witness(
    &self.compiled_circuit,
    sources,
    public_inputs,
    private_inputs,
)?;

// TODO: modify the public output part of the witness

let witness = generated_witness.all_witness.to_kimchi_witness();

// verify the witness
self.index
    .verify(&witness, &generated_witness.full_public_inputs)
    .unwrap();

// create proof
let proof =
    ProverProof::create::<BaseSponge, ScalarSponge>(&GROUP_MAP, witness, &[], &self.index)
        .into_diagnostic()
        .wrap_err("kimchi: could not create a proof with the given inputs")?;

kimchi::verifier::verify::<Curve, BaseSponge, ScalarSponge, OpeningProof<Curve>>(
    &GROUP_MAP,
    &self.index,
    &proof,
    &full_public_inputs,
)
.into_diagnostic()
.wrap_err("kimchi: failed to verify the proof")    
```